[Bug]: Version `29.2.1` breaks default imports

### Version

29.2.1

### Steps to reproduce

1. Clone https://github.com/flovouin/ts-jest-default-imports-bug/.
2. Run `npm i && npm run build`. TypeScript compilation should succeed.
3. Run `npm test`.

### Expected behavior

`npm test` should pass with no error.
You can install `ts-jest` version `29.2.0` instead and see it succeed.

### Actual behavior

```
$ npm test

> test@0.1.0 test
> NODE_OPTIONS="$NODE_OPTIONS --experimental-vm-modules --no-warnings=ExperimentalWarning" jest

 FAIL  src/index.spec.ts
  ● Test suite failed to run

    src/index.ts:2:28 - error TS2351: This expression is not constructable.
      Type 'typeof import("[..]/node_modules/i18next-fs-backend/cjs/index")' has no construct signatures.

    2 export const backend = new Backend();
                                 ~~~~~~~

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.38 s
Ran all test suites.
```

For what it's worth, even though I wasn't expecting to see the CJS version imported, `node_modules/i18next-fs-backend/cjs/index` does exist and export a default class.

### Debug log

[ts-jest.log](https://github.com/user-attachments/files/16345194/ts-jest.log)


### Additional context

_No response_

### Environment

```shell
System:
    OS: macOS 14.5
    CPU: (10) arm64 Apple M1 Max
  Binaries:
    Node: 20.12.1 - ~/.asdf/installs/nodejs/20.12.1/bin/node
    npm: 10.8.2 - ~/.asdf/plugins/nodejs/shims/npm
  npmPackages:
    jest: ^29.7.0 => 29.7.0
```

--- a/e2e/__tests__/native-esm-ts.test.ts
+++ b/e2e/__tests__/native-esm-ts.test.ts
@@ -1,4 +1,11 @@
+import path from 'path'
+
 import { json as runWithJson } from '../run-jest'
+import { runNpmInstall } from '../utils'
+
+beforeAll(() => {
+  runNpmInstall(path.join(__dirname, '..', 'native-esm-ts'))
+})
 
 test('runs TS test with native ESM', () => {
   const { exitCode, json } = runWithJson('native-esm-ts', [], {
@@ -6,6 +13,6 @@ test('runs TS test with native ESM', () => {
   })
 
   expect(exitCode).toBe(0)
-  expect(json.numTotalTests).toBe(4)
-  expect(json.numPassedTests).toBe(4)
+  expect(json.numTotalTests).toBe(6)
+  expect(json.numPassedTests).toBe(6)
 })

--- a/e2e/native-esm-ts/__tests__/third-party.spec.ts
+++ b/e2e/native-esm-ts/__tests__/third-party.spec.ts
@@ -1,0 +1,16 @@
+import path from 'node:path'
+
+import { it, describe, expect } from '@jest/globals'
+
+import { fileName } from '../third-parties/dirname-filename-esm.js'
+import { backend } from '../third-parties/i18next-fs-backend.js'
+
+describe('Third parties', () => {
+  it('should work with i18next-fs-backend', () => {
+    expect(backend).toBeDefined()
+  })
+
+  it('should work with dirname-filename-esm', () => {
+    expect(path.basename(fileName(import.meta))).toBe('third-party.spec.ts')
+  })
+})

--- a/e2e/native-esm-ts/package.json
+++ b/e2e/native-esm-ts/package.json
@@ -1,6 +1,8 @@
 {
   "type": "module",
   "devDependencies": {
-    "@jest/globals": "^29.7.0"
+    "@jest/globals": "^29.7.0",
+    "dirname-filename-esm": "^1.1.2",
+    "i18next-fs-backend": "^2.3.2"
   }
 }

--- a/e2e/native-esm-ts/third-parties/dirname-filename-esm.ts
+++ b/e2e/native-esm-ts/third-parties/dirname-filename-esm.ts
@@ -1,0 +1,3 @@
+import { filename } from 'dirname-filename-esm'
+
+export { filename as fileName }

--- a/e2e/native-esm-ts/third-parties/i18next-fs-backend.ts
+++ b/e2e/native-esm-ts/third-parties/i18next-fs-backend.ts
@@ -1,0 +1,3 @@
+import { default as Backend } from 'i18next-fs-backend'
+
+export const backend = new Backend()

--- a/src/legacy/compiler/ts-compiler.spec.ts
+++ b/src/legacy/compiler/ts-compiler.spec.ts
@@ -200,7 +200,7 @@ describe('TsCompiler', () => {
           useESM: true,
           supportsStaticESM: true,
           moduleValue: 'NodeNext',
-          expectedModule: ts.ModuleKind.NodeNext,
+          expectedModule: ts.ModuleKind.ESNext,
           expectedEsModuleInterop: true,
         },
         {
@@ -252,6 +252,7 @@ describe('TsCompiler', () => {
 
           expect(usedCompilerOptions.module).toBe(expectedModule)
           expect(usedCompilerOptions.esModuleInterop).toBe(expectedEsModuleInterop)
+          expect(usedCompilerOptions.moduleResolution).toBe(ts.ModuleResolutionKind.Node10)
           expect(output).toEqual({
             code: updateOutput(jsOutput, fileName, sourceMap),
             diagnostics: [],

--- a/src/legacy/compiler/ts-compiler.ts
+++ b/src/legacy/compiler/ts-compiler.ts
@@ -147,31 +147,27 @@ export class TsCompiler implements TsCompilerInstance {
   }
 
   private fixupCompilerOptionsForModuleKind(compilerOptions: CompilerOptions, isEsm: boolean): CompilerOptions {
+    const moduleResolution = this._ts.ModuleResolutionKind.Node10
     if (!isEsm) {
-      const moduleKind = compilerOptions.module ?? this._ts.ModuleKind.CommonJS
-      const moduleKindValue = [
-        this._ts.ModuleKind.CommonJS,
-        this._ts.ModuleKind.Node16,
-        this._ts.ModuleKind.NodeNext,
-      ].includes(moduleKind)
-        ? moduleKind
-        : this._ts.ModuleKind.CommonJS
-
       return {
         ...compilerOptions,
-        module: moduleKindValue,
+        module: this._ts.ModuleKind.CommonJS,
+        moduleResolution,
       }
     }
 
-    const moduleKind = compilerOptions.module ?? this._ts.ModuleKind.ESNext
-    const esModuleInterop = [this._ts.ModuleKind.Node16, this._ts.ModuleKind.NodeNext].includes(moduleKind)
-      ? true
-      : compilerOptions.esModuleInterop
+    let moduleKind = compilerOptions.module ?? this._ts.ModuleKind.ESNext
+    let esModuleInterop = compilerOptions.esModuleInterop
+    if ([this._ts.ModuleKind.Node16, this._ts.ModuleKind.NodeNext].includes(moduleKind)) {
+      esModuleInterop = true
+      moduleKind = this._ts.ModuleKind.ESNext
+    }
 
     return {
       ...compilerOptions,
       module: moduleKind,
-      esModuleInterop: esModuleInterop ?? false,
+      esModuleInterop,
+      moduleResolution,
     }
   }
 

